snap/quota/resources.go

```diff
@@ -104,15 +149,46 @@ func (qr *Resources) Change(newLimits Resources) error {
 	if newLimits.Memory != nil {
 		qr.Memory = newLimits.Memory
 	}
+	if newLimits.Cpu != nil {
+		if qr.Cpu == nil {
+			qr.Cpu = newLimits.Cpu
+		} else {
+			// update count/percentage as one unit
+			if newLimits.Cpu.Count != 0 || newLimits.Cpu.Percentage != 0 {
+				qr.Cpu.Count = newLimits.Cpu.Count
+				qr.Cpu.Percentage = newLimits.Cpu.Percentage
+			}
+
+			// update allowed cpus as one unit
+			if len(newLimits.Cpu.AllowedCpus) != 0 {
+				qr.Cpu.AllowedCpus = newLimits.Cpu.AllowedCpus
+			}
+		}
+	}
+	if newLimits.Thread != nil {
+		qr.Thread = newLimits.Thread
+	}
 	return nil
 }
 
-func NewResources(memoryLimit quantity.Size) Resources {
+func NewResources(memoryLimit quantity.Size, cpuCount int, cpuPercentage int, allowedCpus []int, threadLimit int) Resources {
```

Comment:
Looks a bit odd now, and it's hard to tell wherther the 2nd field is CPU count or maybe a thread limit since both are ints.

A rough proposal could be something like (bit Java like, but maybe not that bad):

```
func NewResources() *Resources { &Resources{} }

func (r *Resources) WithCPUCount(count int) *Resources {
    if r.CPU == nil {
        r.CPU = &ResourceCPU{}
    }
    r.CPU.Count = count
    return r
}

func (r *Resources) WithAllowedCPUs(allowedCPUs []int) *Resource {
    if r.CPU == nil {
        r.CPU = &ResourceCPU{}
    }
    r.CPU.AllowedCPUs = allowedCPUs
    return r
}
func (r *Resources) WithMemory(limit int) *Resources {
    r.Memory = &ResourceMemory { Limit: limit }
    return r
}
...
```

Then in the places where it's used, it's quite clear what is being set:

```
r := NewResources().WithAllowedCPUs([]int{1,2}).WithMemory(1024*1024*quantity.KiB)
```

If you don't want to chain the calls, you can skip returning `*Resource`.

Reply:
Good point about the interface for Resources, definitely not that userfriendly. I think it's a good suggestion to consider a builder pattern in this case. I'll take a look at it